resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/tpr/TPRCrudCallsImpl.scala

```diff
@@ -0,0 +1,320 @@
+    response.body().close()
+    logInfo(s"Successfully deleted resource $tprObjectName")
+  }
+
+  override def getJobObject(name: String): SparkJobState = {
+    val requestSegments = Seq(
+      "apis", TPR_API_GROUP, TPR_API_VERSION, "namespaces", namespace, "sparkjobs", name)
+    val url = generateHttpUrl(requestSegments)
+
+    val request = completeRequest(new Request.Builder()
+      .get()
+      .url(url))
+
+    logDebug(s"Get Request: $request")
+    val response = httpClient.newCall(request).execute()
+    completeRequestWithExceptionIfNotSuccessful(
+      "get",
+      response,
+      Option(Seq(name, response.message()))
+    )
+
+    logInfo(s"Successfully retrieved resource $name")
+    read[SparkJobState](response.body().string())
+  }
+
+  override def updateJobObject(name: String, value: String, fieldPath: String): Unit = {
+    val payload = List(
+      ("op" -> "replace") ~ ("path" -> fieldPath) ~ ("value" -> value))
+    val requestBody =
+      RequestBody.create(
+        MediaType.parse("application/json-patch+json"),
+        compact(render(payload)))
+
+    val requestSegments = Seq(
+      "apis", TPR_API_GROUP, TPR_API_VERSION, "namespaces", namespace, "sparkjobs", name)
+    val url = generateHttpUrl(requestSegments)
+
+    val request = completeRequest(new Request.Builder()
+      .patch(requestBody)
+      .url(url))
+
+    logDebug(s"Update Request: $request")
+    val response = httpClient.newCall(request).execute()
+    completeRequestWithExceptionIfNotSuccessful(
+      "patch",
+      response,
+      Option(Seq(name, response.message(), compact(render(payload))))
+    )
+
+    response.body().close()
+    logDebug(s"Successfully patched resource $name.")
+  }
+
+/**
+ * This method has an helper method that blocks to watch the object.
+ * The future is completed on a Delete event or source exhaustion.
+ * This method also relies on the assumption of one sparkjob per namespace
+ */
+  override def watchJobObject(): Future[WatchObject] = {
+    val watchClient = httpClient.newBuilder().readTimeout(0, TimeUnit.MILLISECONDS).build()
+    val requestSegments = Seq(
+      "apis", TPR_API_GROUP, TPR_API_VERSION, "namespaces", namespace, "sparkjobs")
+    val url = generateHttpUrl(requestSegments, Seq(("watch", "true")))
+
+    val request = completeRequest(new Request.Builder()
+      .get()
+      .url(url))
+
+    logDebug(s"Watch Request: $request")
+    val resp = watchClient.newCall(request).execute()
+    completeRequestWithExceptionIfNotSuccessful(
+      "start watch on",
+      resp,
+      Option(Seq(resp.code().toString, resp.message())))
+
+    logInfo(s"Starting watch on jobResource")
+    watchJobObjectUtil(resp)
+  }
+
+/**
+ * This method has a blocking call - wait on SSE - inside it.
+ * However it is sent off in a new thread
+ */
+  private def watchJobObjectUtil(response: Response): Future[WatchObject] = {
+    @volatile var wo: WatchObject = null
+    watchSource = response.body().source()
```

Comment:
Does the fact that this variable is not localized to this method mean that when we watch multiple objects we can lose the watch source? What about if two callers concurrently invoke this method?